freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingService.java

```diff
@@ -0,0 +1,122 @@
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.FormFactor;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceMetadataType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.DownscaleRequest;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.DownscaleResponse;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.ScalingPath;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.UpscaleRequest;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.UpscaleResponse;
+import com.sequenceiq.freeipa.api.v1.operation.model.OperationType;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Operation;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.flow.freeipa.downscale.DownscaleFlowEvent;
+import com.sequenceiq.freeipa.flow.freeipa.downscale.event.DownscaleEvent;
+import com.sequenceiq.freeipa.flow.freeipa.upscale.UpscaleFlowEvent;
+import com.sequenceiq.freeipa.flow.freeipa.upscale.event.UpscaleEvent;
+import com.sequenceiq.freeipa.service.freeipa.flow.FreeIpaFlowManager;
+import com.sequenceiq.freeipa.service.operation.OperationService;
+
+@Service
+public class FreeIpaScalingService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaScalingService.class);
+
+    @Inject
+    private OperationService operationService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaFlowManager flowManager;
+
+    @Inject
+    private FreeIpaScalingValidationService validationService;
+
+    public UpscaleResponse upscale(String accountId, UpscaleRequest request) {
+        Stack stack = getStack(accountId, request.getEnvironmentCrn());
+        Set<InstanceMetaData> allInstances = stack.getNotDeletedInstanceMetaDataSet();
+        FormFactor originalFormFactor = FormFactor.getByInstanceCount(allInstances.size());
+        validationService.validateStackForUpscale(allInstances, stack, new ScalingPath(originalFormFactor, request.getTargetFormFactor()));
+        return triggerUpscale(request, stack, originalFormFactor);
+    }
+
+    public DownscaleResponse downscale(String accountId, DownscaleRequest request) {
+        Stack stack = getStack(accountId, request.getEnvironmentCrn());
+        Set<InstanceMetaData> allInstances = stack.getNotDeletedInstanceMetaDataSet();
+        FormFactor originalFormFactor = FormFactor.getByInstanceCount(allInstances.size());
+        validationService.validateStackForDownscale(allInstances, stack, new ScalingPath(originalFormFactor, request.getTargetFormFactor()));
+        return triggerDownscale(request, stack, originalFormFactor);
+    }
+
+    private Stack getStack(String accountId, String envCrn) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(envCrn, accountId);
+        MDCBuilder.buildMdcContext(stack);
+        return stack;
+    }
```

Comment:
it is worth to move this method into a common service since this logic is used in lot of classes across freeipa module (CleanupService, DiagnosticsService, RepairInstancesService, etc,)